pages/_app/index.jsx

```diff
@@ -35,19 +32,24 @@ class App extends NextApp {
 
     event.preventDefault()
     const route = new Route(url.pathname)
-    this.props.router.push(route.href, route.as)
+    await this.props.router.push(route.href, route.as)
   }
 
   render() {
-    const { Component, pageProps, initialStoreData, router } = this.props
+    const { Component, pageProps, store, router } = this.props
 
     const { dataProvider, activity } = new Route(router.asPath)
 
+    // needs to be done here because getInitialProps can be called before
+    // handleNavigation (router.push) is done
+    store.dataProvider = dataProvider
+    store.activity = activity
+
     return (
-      <GlobalProvider initialData={initialStoreData}>
+      <GlobalProvider store={store}>
         <Application
-          dataProvider={dataProvider}
-          activity={activity}
+          dataProvider={store.dataProvider}
+          activity={store.activity}
```

Comment:
I wonder if this can be calculated somehow based on PageComponent but not for this moment.